components/places/src/bookmark_sync/store.rs

```diff
@@ -0,0 +1,1233 @@
+        sql_support::each_chunk(&guids, |chunk, _| -> Result<()> {
+            self.db.execute(
+                &format!(
+                    "UPDATE itemsToUpload SET
+                       uploadedAt = {uploaded_at}
+                     WHERE guid IN ({values})",
+                    uploaded_at = uploaded_at.as_millis(),
+                    values = sql_support::repeat_sql_values(chunk.len())
+                ),
+                chunk,
+            )?;
+            Ok(())
+        })?;
+
+        // Fast-forward the last sync time, so that we don't download the
+        // records we just uploaded on the next sync.
+        put_meta(
+            self.db,
+            LAST_SYNC_META_KEY,
+            &(uploaded_at.as_millis() as i64),
+        )?;
+
+        // Clean up.
+        self.db.execute_batch("DELETE FROM itemsToUpload")?;
+
+        Ok(())
+    }
+}
+
+impl<'a> Store for BookmarksStore<'a> {
+    #[inline]
+    fn collection_name(&self) -> &'static str {
+        "bookmarks"
+    }
+
+    fn apply_incoming(
+        &self,
+        inbound: IncomingChangeset,
+        incoming_telemetry: &mut telemetry::EngineIncoming,
+    ) -> result::Result<OutgoingChangeset, failure::Error> {
+        use dogear::Store;
+
+        // Stage all incoming items.
+        let timestamp = self.stage_incoming(inbound, incoming_telemetry)?;
+
+        // write the timestamp now, so if we are interrupted merging or
+        // creating outgoing changesets we don't need to re-download the same
+        // records.
+        put_meta(self.db, LAST_SYNC_META_KEY, &(timestamp.as_millis() as i64))?;
+
+        // Merge and stage outgoing items.
+        self.merge_with_driver(&Driver)?;
+
+        let outgoing = self.fetch_outgoing_records(timestamp)?;
+        Ok(outgoing)
+    }
+
+    fn sync_finished(
+        &self,
+        new_timestamp: ServerTimestamp,
+        records_synced: &[String],
+    ) -> result::Result<(), failure::Error> {
+        let tx = self.db.unchecked_transaction()?;
+        let result = self.push_synced_items(new_timestamp, records_synced);
+        match result {
+            Ok(_) => tx.commit()?,
+            Err(_) => tx.rollback()?,
+        }
+        result?;
+        Ok(())
+    }
+
+    fn get_collection_request(&self) -> result::Result<CollectionRequest, failure::Error> {
+        let since = get_meta::<i64>(self.db, LAST_SYNC_META_KEY)?
+            .map(|millis| ServerTimestamp(millis as f64 / 1000.0))
+            .unwrap_or_default();
+        Ok(CollectionRequest::new(self.collection_name())
+            .full()
+            .newer_than(since))
+    }
+
+    fn reset(&self) -> result::Result<(), failure::Error> {
+        unimplemented!("TODO: Wipe staged items and reset sync statuses");
+    }
+
+    fn wipe(&self) -> result::Result<(), failure::Error> {
+        log::warn!("not implemented");
+        Ok(())
+    }
+}
+
+fn type_to_kind<'a>(typ: &'a str, url: UrlOrPlaceId<'a>) -> TypeToKind<'a> {
```

Comment:
I learned this trick from @thomcc's `sql_support`. Maybe it needs a better name, though? `type_to_kind_fragment`?